### PR TITLE
feat(upgrade-checker):  implement extra info (tag, field)

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn-manager/pull/1891#issuecomment-1588645215

https://github.com/longhorn/longhorn/issues/5235

### CODE REVIEW

**Review by innobead — DISMISSED** [review]
LGTM, but pending for testing.

**Comment by innobead** [comment]
cc @PhanLe1010 can we still leverage extra info to send the field data? Should we need to have. An extra design for field data in extra info?

**Comment by PhanLe1010** [comment]
> cc @PhanLe1010 can we still leverage extra info to send the field data? Should we need to have. An extra design for field data in extra info?

I think we will extends the the extraInfo have 2 types:

1. ExtraTagInfo: each value here will be tag in InfluxDB. This must be string
2. ExtraFieldInfo: each value here will be a field in InfluxDB. This can be int. float, string,...

Then support backward compatibility by consider extraInfo as ExtraTagInfo. What do you think?

**Comment by PhanLe1010** [comment]
Thanks @c3y1huang for updating the PR to reflecting this discussion https://github.com/longhorn/longhorn-manager/pull/1975#issuecomment-1590141500. I am standing up a dev env for upgrade responder to test the whole flow. Will return with feedback soon

File: controller/setting_controller.go
Comment (innobead) [inline controller/setting_controller.go:1390]: if failed to convert, do we still need to collect that value? probably just exclude to prevent any bias or potential error?

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Two comments need to resolve.

File: controller/setting_controller.go
Comment (innobead) [reply controller/setting_controller.go]: cc @derekbit @shuo-wu

File: controller/setting_controller.go
Comment (innobead) [inline controller/setting_controller.go:1528]: `totalVolumeActualSize / (volumeCount - v2VolumeCount)`

File: controller/setting_controller.go
Comment (innobead) [inline controller/setting_controller.go:1465]: Only `actual size` is unable to supported by v2 data engine, so `volume.spec.size` still needs to include.

File: controller/setting_controller.go
Comment (innobead) [inline controller/setting_controller.go:1451]: Can you have newline to organize the code better like below?

```go
	var totalVolumeSize int
	var totalVolumeActualSize int
	var totalVolumeNumOfReplicas int

	newStruct := func() map[util.StructName]int { 
        return make(map[util.StructName]int, volumeCount) 
    }

    accessModeCountStruct := newStruct()
	dataLocalityCountStruct := newStruct()
	frontendCountStruct := newStruct()
	replicaAutoBalanceCountStruct := newStruct()
	
    for _, volume := range volumesRO {
```

**Review by innobead — CHANGES_REQUESTED** [review]
In genreal, LGTM. Just a few feedback.

**Review by innobead — DISMISSED** [review]

**Comment by innobead** [comment]
After @PhanLe1010 review, we can merge this.

File: controller/setting_controller.go
Comment (c3y1huang) [reply controller/setting_controller.go:1249]: We are collecting the Longhorn namespace UID to count the number of adopted clusters.

Comment (PhanLe1010) [reply controller/setting_controller.go:1249]: Thank you very much

**Comment by PhanLe1010** [comment]
We are creating tag for string and bool setting type:
```
longhorn_setting_allow_recurring_job_while_volume_detached
longhorn_setting_allow_volume_creation_with_degraded_availability
longhorn_setting_auto_cleanup_system_generated_snapshot
longhorn_setting_auto_delete_pod_when_volume_detached_unexpectedly
longhorn_setting_auto_salvage
longhorn_setting_backup_compression_method
longhorn_setting_backup_target
longhorn_setting_crd_api_version
longhorn_setting_create_default_disk_labeled_nodes
longhorn_setting_default_data_locality
longhorn_setting_disable_revision_counter
longhorn_setting_disable_scheduling_on_cordoned_node
longhorn_setting_fast_replica_rebuild_enabled
longhorn_setting_guaranteed_instance_manager_cpu
longhorn_setting_kubernetes_cluster_autoscaler_enabled
longhorn_setting_node_down_pod_deletion_policy
longhorn_setting_node_drain_policy
longhorn_setting_offline_replica_rebuilding
longhorn_setting_orphan_auto_deletion
longhorn_setting_priority_class
longhorn_setting_registry_secret
longhorn_setting_remove_snapshots_during_filesystem_trim
longhorn_setting_replica_auto_balance
longhorn_setting_replica_soft_anti_affinity
longhorn_setting_replica_zone_soft_anti_affinity
longhorn_setting_restore_volume_recurring_jobs
longhorn_setting_snapshot_data_integrity
longhorn_setting_snapshot_data_integrity_cronjob
longhorn_setting_snapshot_data_integrity_immediate_check_after_snapshot_creation
longhorn_setting_storage_network
longhorn_setting_system_managed_components_node_selector
longhorn_setting_system_managed_pods_image_pull_policy
longhorn_setting_taint_toleration
longhorn_setting_v2_data_engine
```

Could you elaborate the reason for using tag instead of field? Thank you!

**Comment by c3y1huang** [comment]
> We are creating tags for string and bool setting types:
> 
> ```
> longhorn_setting_allow_recurring_job_while_volume_detached
> longhorn_setting_allow_volume_creation_with_degraded_availability
> longhorn_setting_auto_cleanup_system_generated_snapshot
> longhorn_setting_auto_delete_pod_when_volume_detached_unexpectedly
> longhorn_setting_auto_salvage
> longhorn_setting_backup_compression_method
> longhorn_setting_backup_target
> longhorn_setting_crd_api_version
> longhorn_setting_create_default_disk_labeled_nodes
> longhorn_setting_default_data_locality
> longhorn_setting_disable_revision_counter
> longhorn_setting_disable_scheduling_on_cordoned_node
> longhorn_setting_fast_replica_rebuild_enabled
> longhorn_setting_guaranteed_instance_manager_cpu
> longhorn_setting_kubernetes_cluster_autoscaler_enabled
> longhorn_setting_node_down_pod_deletion_policy
> longhorn_setting_node_drain_policy
> longhorn_setting_offline_replica_rebuilding
> longhorn_setting_orphan_auto_deletion
> longhorn_setting_priority_class
> longhorn_setting_registry_secret
> longhorn_setting_remove_snapshots_during_filesystem_trim
> longhorn_setting_replica_auto_balance
> longhorn_setting_replica_soft_anti_affinity
> longhorn_setting_replica_zone_soft_anti_affinity
> longhorn_setting_restore_volume_recurring_jobs
> longhorn_setting_snapshot_data_integrity
> longhorn_setting_snapshot_data_integrity_cronjob
> longhorn_setting_snapshot_data_integrity_immediate_check_after_snapshot_creation
> longhorn_setting_storage_network
> longhorn_setting_system_managed_components_node_selector
> longhorn_setting_system_managed_pods_image_pull_policy
> longhorn_setting_taint_toleration
> longhorn_setting_v2_data_engine
> ```
> 
> Could you elaborate the reason for using tag instead of field? Thank you!

We are using the tag for non-numeric settings to display the counted values better in the Grafana panels. And we don't need to do the calculation.

**Comment by PhanLe1010** [comment]
When using tag, I think the set of possible values should be small and predictable to avoid [the high series cardinality issue](https://docs.influxdata.com/influxdb/cloud/write-data/best-practices/resolve-high-cardinality/#learn-the-causes-of-high-series-cardinality) which might lead to high memory usage and crash our influxDB. 

Some settings that could potentially have a big set of values are:
```
longhorn_setting_guaranteed_instance_manager_cpu
longhorn_setting_priority_class
longhorn_setting_registry_secret
longhorn_setting_storage_network
longhorn_setting_system_managed_components_node_selector
longhorn_setting_taint_toleration
```

Should we use field for these setting?

**Comment by c3y1huang** [comment]
> When using tag, I think the set of possible values should be small and predictable to avoid [the high series cardinality issue](https://docs.influxdata.com/influxdb/cloud/write-data/best-practices/resolve-high-cardinality/#learn-the-causes-of-high-series-cardinality) which might lead to high memory usage and crash our influxDB.
> 
> Some settings that could potentially have a big set of values are:
> 
> ```
> longhorn_setting_guaranteed_instance_manager_cpu
> longhorn_setting_priority_class
> longhorn_setting_registry_secret
> longhorn_setting_storage_network
> longhorn_setting_system_managed_components_node_selector
> longhorn_setting_taint_toleration
> ```
> 
> Should we use field for these setting?
> 
> Btw, is it correct that we just need to know if the users use these features (settings)? If yes, should we use bool value instead of storing the exact setting?

Below are [collected as bool](https://github.com/longhorn/longhorn-manager/blob/master/controller/setting_controller.go#L1300-L1306) already:
-  longhorn_setting_priority_class
-  longhorn_setting_registry_secret
- longhorn_setting_storage_network
- longhorn_setting_system_managed_components_node_selector
- longhorn_setting_taint_toleration

The setting values are collected based on the setting definition. I have updated the `longhorn_setting_guaranteed_instance_manager_cpu` setting type to int.

**Comment by PhanLe1010** [comment]
Thank you! The PR LGTM. I am doing some small testing on gravatar side and will update here soon

**Review by innobead — APPROVED** [review]

**Comment by innobead** [comment]
@mergify backport v1.5.x

**Pull request merged**
